corehq/apps/app_manager/util.py

```diff
@@ -318,28 +321,52 @@ def get_all_case_properties(app):
     return get_case_properties(app, app.get_case_types(), defaults=('name',))
 
 
-def get_casedb_schema(app):
-    """Get case database schema definition
+def get_casedb_schema(form):
+    """Get case database schema definition for vellum to display as an external data source.
 
     This lists all case types and their properties for the given app.
     """
+    app = form.get_app()
+    base_case_type = form.get_module().case_type
     case_types = app.get_case_types()
     per_type_defaults = get_per_type_defaults(app.domain, case_types)
     builder = ParentCasePropertyBuilder(app, ['case_name'], per_type_defaults)
-    related = builder.get_parent_type_map(case_types)
+    related = builder.get_parent_type_map(case_types, allow_multiple_parents=True)
     map = builder.get_case_property_map(case_types, include_parent_properties=False)
+
+    if base_case_type:
+        # Generate hierarchy of case types, represented as a list of lists of strings:
+        # [[base_case_type], [parent_type1, parent_type2...], [grandparent_type1, grandparent_type2...]]
+        # Vellum case management only supports three levels
+        generation_names = ['case', 'parent', 'grandparent']
+        generations = [[] for g in generation_names]
+
+        def _add_ancestors(ctype, generation):
+            if generation < len(generation_names):
+                generations[generation].append(ctype)
+                for parent in related.get(ctype, {}).get('parent', []):
+                    _add_ancestors(parent, generation + 1)
+
+        _add_ancestors(base_case_type, 0)
+
+        # Remove any duplicate types or empty generations
+        generations = [set(g) for g in generations if len(g)]
```

Comment:
I think you can get rid of this if you use `generations = [set() for g in generation_names]` above